src/Microsoft.ML.Core/ComponentModel/ComponentCatalog.cs

```diff
@@ -414,21 +414,78 @@ internal ComponentInfo(Type interfaceType, string kind, Type argumentType, TlcMo
             ctor = null;
             create = null;
             requireEnvironment = false;
+            bool requireEnvironmentCtor = false;
+            bool requireEnvironmentCreate = false;
             var parmTypesWithEnv = Utils.Concat(new Type[1] { typeof(IHostEnvironment) }, parmTypes);
+
             if (Utils.Size(parmTypes) == 0 && (getter = FindInstanceGetter(instType, loaderType)) != null)
                 return true;
-            if (instType.IsAssignableFrom(loaderType) && (ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypes ?? Type.EmptyTypes, null)) != null)
-                return true;
-            if (instType.IsAssignableFrom(loaderType) && (ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypesWithEnv ?? Type.EmptyTypes, null)) != null)
+
+            // Find both 'ctor' and 'create' methods if available
+            if (instType.IsAssignableFrom(loaderType))
             {
-                requireEnvironment = true;
-                return true;
+                ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypes ?? Type.EmptyTypes, null);
+                if (ctor == null)
+                {
+                    ctor = loaderType.GetConstructor(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic, null, parmTypesWithEnv ?? Type.EmptyTypes, null);
+                    if (ctor != null)
+                        requireEnvironmentCtor = true;
+                }
             }
-            if ((create = FindCreateMethod(instType, loaderType, parmTypes ?? Type.EmptyTypes)) != null)
+
+            create = FindCreateMethod(instType, loaderType, parmTypes ?? Type.EmptyTypes);
+            if (create == null)
+            {
+                create = FindCreateMethod(instType, loaderType, parmTypesWithEnv ?? Type.EmptyTypes);
+                if (create != null)
+                    requireEnvironmentCreate = true;
+            }
+
+            // If both 'ctor' and 'create' methods were found
+            // Choose the one that is public, if both are non-public choose the one that isn't private
+            // if they have the same visibility, then throw an exception, since this shouldn't happen.
+            if(ctor != null && create != null)
+            {
+                if (ctor.IsPublic == create.IsPublic)
+                {
+                    if(ctor.IsPublic || ctor.IsPrivate == create.IsPrivate)
+                    {
+                        throw Contracts.Except($"Can't load type {instType}, because it has both create and constructor methods with the same visibility. Please open an issue for this to be fixed.");
```

Comment:
I didn't know what text to put in the Exception. Please let me know if I should change it.

Reply:
I would change "please open an issue..." to something like: "Please indicate which one should be used by changing either the signature or the visibility of one of them".

---
In reply to: [347643643](https://github.com/dotnet/machinelearning/pull/4485#discussion_r347643643) [](ancestors = 347643643)